### PR TITLE
[최윤혁] TranscriptServiceImplTest 유닛테스트 작성

### PR DESCRIPTION
**Pull request by cyh789**
# 개요
TranscriptServiceImplTest 유닛테스트 코드 작성

# 작업사항
Mockito를 이용하여 테스트 구현

### CODE REVIEW

File: src/test/java/com/flab/service/TranscriptServiceImplTest.java [hunk 1]
```diff
@@ -82,29 +85,146 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
     @Test
     @DisplayName("getAverageScore()에서 studentRepository.getStudent()를 한 번, scoreRepository.getScore()를 student의 course 개수만큼 호출한다.")
```
Comment (f-lab-steve) [inline src/test/java/com/flab/service/TranscriptServiceImplTest.java:86]: 보통 이런 로직을 테스트할 때는, edge case들을 위주로 테스트해봅니다. 

"어떤 학생이 특정 과목들을 수강한다고 했을 때, 시험을 꼭 쳐서 점수가 입력되어 있다"라는 가정을 할 수도 있고.
"시험을 아직 일부만 치뤘지만, 중간 평균점수를 구할 수 있어야 한다"라는 가정을 할 수도 있습니다. 

구현은 일단 1번째 방법 입니다

그래서 테스트 케이스가 2개로 나뉩니다.
1. N개의 과목중 0 <= X < N개의 스코어가 존재하는 경우 -> exception이 잘 throw되는지 체크 (별도의 테스트 케이스가 있음)
2. N개의 과목중 N개의 스코어가 존재하는 경우 -> 현재의 테스트 케이스

그 중 N개의 과목을 한번이라도 테스트 했다면, 중복 테스트이므로 굳이 2번의 테스트는 필요 없어 보입니다. trey 혹은 cyh케이스만 남기시면 됩니다.

File: src/test/java/com/flab/service/TranscriptServiceImplTest.java [hunk 2]
```diff
@@ -82,29 +85,146 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
...
-        throw new UnsupportedOperationException("Not implemented yet");
+        // given
+        final int studentID = 1;
+        final Student trey = new Student().setId(studentID).setName("Trey").setMajor("Computer Engineering")
+                .setCourses(List.of(KOREAN, ENGLISH, MATH, SCIENCE));
+
+        Mockito.when(studentRepository.getStudent(1))
+                .thenReturn(Optional.of(trey));
+
+        Mockito.when(scoreRepository.getScore(studentID, 1))
```
Comment (f-lab-steve) [inline src/test/java/com/flab/service/TranscriptServiceImplTest.java:143]: 이런식으로 구체적인 (정확한) parameter를 지정하는 것도 가능하지만, [ArgumentMatchers](https://site.mockito.org/javadoc/current/org/mockito/ArgumentMatchers.html)라는게 존재합니다. 해당 기능을 이용하면 어떤 값이 들어오더라도 특정 값을 반환할 수 있도록 mocking이 가능합니다.

File: src/test/java/com/flab/service/TranscriptServiceImplTest.java [hunk 3]
```diff
@@ -82,29 +85,146 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
...
```
Comment (f-lab-steve) [inline src/test/java/com/flab/service/TranscriptServiceImplTest.java:96]: 이렇게 magic number같이 1, 2, 3, 4를 raw value로 넘겨버리면 나중에 유지보수가 어렵습니다. 이건 테스트코드에서도 마찬가지입니다. 위쪽에 사전에 정의된 Course instances (KOREAN, ENGLISH 등)들을 이용하여 값을 참조하거나, `static final int KOREAN_COURSE_ID = 1`과 같이 정의하고 기존 Course instance생성시, 그리고 getScore mocking에 해당 아이디를 사용하는게 좋아보입니다.

File: src/test/java/com/flab/service/TranscriptServiceImplTest.java [hunk 4]
```diff
@@ -82,29 +85,146 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
...
+                .thenReturn(Optional.empty());
+
+        // when & then
+        Assertions.assertThrows(NoSuchScoreException.class, () -> transcriptService.getAverageScore(studentID));
     }
 
     @Test
     @DisplayName("getRankedStudentAsc()를 호출하면, 입력으로 주어진 course를 수강하는 모든 학생들의 리스트를 성적의 내림차순으로 리턴한다.")
     void testGetRankedStudentsAsc_HappyCase_VerifyReturnedValueAndInteractions_Success() {
```
Comment (f-lab-steve) [inline src/test/java/com/flab/service/TranscriptServiceImplTest.java:152]: 여기도 같은 내용을 테스트하는 코드가 중복으로 들어가 있어서 테스트 하는 내용을 파악하는데 어려움이 있는 것 같습니다.

테스트 해야 할 내용은 해당 과목을 수강하는 모든 학생들이 리스트에 포함되어있고, 성적이 내림차순인지입니다.
1. 리스트에 포함 여부를 테스트 해봐야합니다.
1.1 특정 과목을 듣는 학생들을 설정하여 해당 학생들이 최종 결과에 포함되었는지 체크
1.2 학생들의 리스트를 만들 때 특정 과목을 듣지 않도록 설정하여 해당 학생이 최종 결과에 미포함 되었는지 체크
2. 학생들이 내림차순으로 정렬되었는지 체크

>

File: src/test/java/com/flab/service/TranscriptServiceImplTest.java [hunk 5]
```diff
@@ -82,29 +85,146 @@ void testGetAverageScore_StudentNotExist_ThrowNoSuchStudentException_Error() {
...
+        Mockito.when(studentRepository.getAllStudents())
+                .thenReturn(studentsList);
+
+        // then
+        List<Student> rankedStudentsList = transcriptService.getRankedStudentsAsc(1);
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < rankedStudentsList.size(); i++) {
+            sb.append(rankedStudentsList.get(i).getId() + " ");
+        }
+        Assertions.assertEquals(String.valueOf(sb),"3 2 1 ");
```
Comment (f-lab-steve) [inline src/test/java/com/flab/service/TranscriptServiceImplTest.java:209]: 이런식으로 최종 결과물을 가공하여 비교하는 것도 가능하지만, 결과값으로 기대하는 (expected) List<String>을 만들어서 Equals비교도 가능합니다 (Arrays.asList()같은 util을 사용하면 쉽겠습니다)

**Review by f-lab-steve — CHANGES_REQUESTED** [review]
